grails-app/services/org/pih/warehouse/core/IdentifierService.groovy

```diff
@@ -150,13 +153,83 @@ class IdentifierService {
         return identifier.toUpperCase()
     }
 
+    def generateSequentialProductIdentifier(ProductType productType, String sequentialPatternChar, Integer allowedSequences) {
+        // If custom identifier contains more than one sequential part, then throw exception
+        Pattern pattern = Pattern.compile("${sequentialPatternChar}+")
+        Matcher matcher = pattern.matcher(productType.productIdentifierFormat)
+        int count = 0
+        def sequenceFormat = ""
+        while (matcher.find()) {
+            sequenceFormat = matcher.group()
+            count++
+        }
+        if (count > allowedSequences) {
+            throw new IllegalArgumentException("Cannot have more sequence numbers than ${allowedSequences} in the same identifier")
+        }
+
+        def sequenceNumber = productType.getNextSequenceNumber()
+        List identifierFormatComponents
+        List identifierComponents
+
+        // Split custom identifier by sequence format (with keeping sequence inside identifier components)
+        String sequenceDelimiter = "((?<=${sequenceFormat})|(?=${sequenceFormat}))"
+        identifierFormatComponents = productType.productIdentifierFormat.split(sequenceDelimiter)
+        identifierComponents = identifierFormatComponents.collect { String identifierFormatComponent ->
+            if (identifierFormatComponent.contains("0")) {
+                return generateSequenceNumber(sequenceNumber.toString(), identifierFormatComponent)
+            }
+            else {
+                return generateIdentifier(identifierFormatComponent)
+            }
+        }
+
+        return identifierComponents.join("")
+    }
+
+    String generateProductIdentifier(ProductType productType) {
+        // If the product type is null, then generate product code from DEFAULT_PRODUCT_NUMBER_FORMAT
+        if (!productType) {
+            return generateProductIdentifier()
+        }
+
+        // If the product type does not have a custom identifier but has code, then generate sequential product code
+        if (!productType.productIdentifierFormat && productType.code) {
+            Integer sequenceNumber = productType.getNextSequenceNumber()
```

Comment:
Does it buy us anything to move this into a separate method? Easier to unit test?